File: sdks/csharp/Sauron/Queue.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Sauron;

/// <summary>One queued, serialized envelope awaiting delivery, plus its FIFO id and backing file.</summary>
internal sealed class QueueEntry
{
    public long Id { get; }
    public byte[] Payload { get; }

    /// <summary>Backing file when disk persistence is on; null for in-memory entries.</summary>
    public string? File { get; }

    public QueueEntry(long id, byte[] payload, string? file)
    {
        Id = id;
        Payload = payload;
        File = file;
    }
}

/// <summary>
/// A bounded, FIFO buffer of pending envelopes — the transient-outage store. It is byte-capped
/// (<c>maxBytes</c>): pushing past the cap drops the oldest entries so memory can't grow without
/// bound. When an <c>offlineDir</c> is supplied, entries are also persisted to disk (one file each,
/// named by a zero-padded sequence for FIFO ordering), reloaded on construction, and deleted once
/// acknowledged — giving at-least-once delivery across process restarts.
/// </summary>
internal sealed class BoundedQueue
{
    private const string Ext = ".env";

    private readonly int _maxBytes;
    private readonly string? _dir;
    private readonly object _gate = new();
    private readonly LinkedList<QueueEntry> _entries = new();
    private long _bytes;
    private long _seq;

    public BoundedQueue(int maxBytes, string? offlineDir)
    {
        _maxBytes = maxBytes < 0 ? 0 : maxBytes;
        _dir = string.IsNullOrEmpty(offlineDir) ? null : offlineDir;

        if (_dir is not null)
        {
            try { Directory.CreateDirectory(_dir); }
            catch { _dir = null; } // if the dir is unusable, fall back to memory-only
            if (_dir is not null)
                LoadFromDisk();
        }
    }

    /// <summary>Total bytes currently retained.</summary>
    public long Bytes { get { lock (_gate) { return _bytes; } } }

    /// <summary>Number of entries currently retained.</summary>
    public int Count { get { lock (_gate) { return _entries.Count; } } }

    /// <summary>Append an envelope payload, persisting it and evicting the oldest over the cap.</summary>
    public void Push(byte[] payload)
    {
        if (payload is null)
            return;

        lock (_gate)
        {
            long id = ++_seq;
            string? file = null;
            if (_dir is not null)
            {
                file = Path.Combine(_dir, id.ToString("D20") + Ext);
                try { System.IO.File.WriteAllBytes(file, payload); }
                catch { file = null; }
            }

            _entries.AddLast(new QueueEntry(id, payload, file));
            _bytes += payload.Length;
            EnforceCap();
        }
    }

    /// <summary>A FIFO snapshot of the current entries (for a drain pass); does not mutate the queue.</summary>
    public IReadOnlyList<QueueEntry> Snapshot()
    {
        lock (_gate) { return _entries.ToList(); }
    }

    /// <summary>Remove a delivered (or permanently-dropped) entry and delete its backing file.</summary>
    public void Ack(QueueEntry entry)
    {
        if (entry is null)
            return;

        lock (_gate)
        {
            var node = _entries.Find(entry);
            if (node is not null)
            {
                _entries.Remove(node);
                _bytes -= entry.Payload.Length;
                if (_bytes < 0) _bytes = 0;
            }
            DeleteFile(entry);
        }
    }

    /// <summary>Remove and return every entry (FIFO), deleting all backing files.</summary>
    public IReadOnlyList<QueueEntry> Drain()
    {
        lock (_gate)
        {
            var all = _entries.ToList();
            foreach (var e in all)
                DeleteFile(e);
            _entries.Clear();
            _bytes = 0;
            return all;
        }
    }

    private void EnforceCap()
    {
        while (_bytes > _maxBytes && _entries.First is { } first)
        {
            _entries.RemoveFirst();
            _bytes -= first.Value.Payload.Length;
            if (_bytes < 0) _bytes = 0;
            DeleteFile(first.Value);
        }
    }

    private static void DeleteFile(QueueEntry entry)
    {
        if (entry.File is null)
            return;
        try { if (System.IO.File.Exists(entry.File)) System.IO.File.Delete(entry.File); }
        catch { /* best-effort */ }
    }

    private void LoadFromDisk()
    {
        if (_dir is null)
            return;

        string[] files;
        try { files = Directory.GetFiles(_dir, "*" + Ext); }
        catch { return; }

        Array.Sort(files, StringComparer.Ordinal); // zero-padded ids sort FIFO

        foreach (var f in files)
        {
            byte[] payload;
            try { payload = System.IO.File.ReadAllBytes(f); }
            catch { continue; }

            long id = ParseId(Path.GetFileNameWithoutExtension(f)) ?? (_seq + 1);
            if (id > _seq) _seq = id;

            _entries.AddLast(new QueueEntry(id, payload, f));
            _bytes += payload.Length;
        }

        EnforceCap();
    }

    private static long? ParseId(string name)
        => long.TryParse(name, out var v) ? v : (long?)null;
}

File: sdks/python/sauron/_queue.py
"""A bounded, thread-safe outbound queue with opt-in disk persistence.

Default behavior is an in-memory FIFO ring bounded by a byte budget
(``max_bytes``): once the pending payload exceeds the budget the oldest entries
are dropped. This gives a server a practical "offline" story — survive a
transient ingest outage without unbounded memory growth.

When ``offline_path`` is set, every pending item is also written to a file under
that directory (one file per item, FIFO-ordered by filename). A fresh instance
over the same directory reloads the pending items on construction, and
:meth:`confirm` deletes the files once a batch has been delivered — giving
at-least-once delivery across process restarts. Disk persistence is off by
default (ephemeral containers shouldn't be forced into filesystem assumptions).
"""

from __future__ import annotations

import json
import os
import threading
import time
from collections import deque
from typing import Any, Deque, Dict, List, Mapping, Optional

# 1 MiB: matches the js/flutter default queue budget.
DEFAULT_MAX_QUEUE_BYTES = 1_048_576

_SUFFIX = ".json"


class QueueEntry:
    """One pending item plus its on-disk bookkeeping."""

    __slots__ = ("item", "size", "filename")

    def __init__(
        self, item: Dict[str, Any], size: int, filename: Optional[str]
    ) -> None:
        self.item = item
        self.size = size
        self.filename = filename


class BoundedQueue:
    """A byte-bounded FIFO of envelope items with optional disk persistence."""

    def __init__(
        self,
        *,
        max_bytes: int = DEFAULT_MAX_QUEUE_BYTES,
        offline_path: Optional[str] = None,
    ) -> None:
        self._max_bytes = max_bytes
        self._offline_path = offline_path
        self._entries: Deque[QueueEntry] = deque()
        self._bytes = 0
        self._seq = 0
        self._lock = threading.Lock()

        if self._offline_path:
            os.makedirs(self._offline_path, exist_ok=True)
            self._reload()

    # -- producer ----------------------------------------------------------

    def push(self, item: Mapping[str, Any]) -> None:
        """Append an item, persisting it (if enabled) and enforcing the cap."""
        item = dict(item)
        payload = _encode(item)
        with self._lock:
            filename = self._persist(payload)
            entry = QueueEntry(item=item, size=len(payload), filename=filename)
            self._entries.append(entry)
            self._bytes += entry.size
            self._evict_over_cap()

    # -- consumer ----------------------------------------------------------

    def drain(self) -> List[QueueEntry]:
        """Remove and return every pending entry (memory only).

        On-disk files are retained until :meth:`confirm` — so a send that fails
        (or a crash mid-flight) leaves them for the next process to recover.
        """
        with self._lock:
            entries = list(self._entries)
            self._entries.clear()
            self._bytes = 0
            return entries

    def confirm(self, entries: List[QueueEntry]) -> None:
        """Delete the persisted files backing ``entries`` (after delivery)."""
        if not self._offline_path:
            return
        for entry in entries:
            self._remove_file(entry.filename)

    def clear(self) -> None:
        """Drop everything from memory and disk (e.g. on a hard disable)."""
        with self._lock:
            entries = list(self._entries)
            self._entries.clear()
            self._bytes = 0
        for entry in entries:
            self._remove_file(entry.filename)

    # -- introspection -----------------------------------------------------

    @property
    def bytes(self) -> int:
        return self._bytes

    def __len__(self) -> int:
        return len(self._entries)

    def snapshot_items(self) -> List[Dict[str, Any]]:
        """The pending items in FIFO order (test/inspection aid)."""
        with self._lock:
            return [dict(e.item) for e in self._entries]

    # -- internals ---------------------------------------------------------

    def _evict_over_cap(self) -> None:
        # Caller holds the lock. Drop oldest until within budget, but always
        # keep at least the newest entry (a lone oversized item still ships).
        while self._bytes > self._max_bytes and len(self._entries) > 1:
            victim = self._entries.popleft()
            self._bytes -= victim.size
            self._remove_file(victim.filename)

    def _persist(self, payload: bytes) -> Optional[str]:
        # Caller holds the lock.
        if not self._offline_path:
            return None
        # A time-then-sequence prefix keeps files FIFO-sortable across restarts.
        self._seq += 1
        filename = "%020d_%08d%s" % (time.time_ns(), self._seq, _SUFFIX)
        path = os.path.join(self._offline_path, filename)
        try:
            with open(path, "wb") as fh:
                fh.write(payload)
        except OSError:
            # Persistence is best-effort; the in-memory copy still ships.
            return None
        return filename

    def _remove_file(self, filename: Optional[str]) -> None:
        if not filename or not self._offline_path:
            return
        try:
            os.remove(os.path.join(self._offline_path, filename))
        except FileNotFoundError:
            pass
        except OSError:
            pass

    def _reload(self) -> None:
        # Caller is __init__ (single-threaded). Recover pending items FIFO.
        try:
            names = sorted(
                n for n in os.listdir(self._offline_path) if n.endswith(_SUFFIX)
            )
        except OSError:
            return
        for name in names:
            path = os.path.join(self._offline_path, name)
            try:
                with open(path, "rb") as fh:
                    payload = fh.read()
                item = json.loads(payload.decode("utf-8"))
            except (OSError, ValueError):
                # A corrupt/partial file can't be recovered — discard it.
                self._remove_file(name)
                continue
            entry = QueueEntry(item=item, size=len(payload), filename=name)
            self._entries.append(entry)
            self._bytes += entry.size
        # Reloaded files may already exceed the budget (e.g. a smaller cap on
        # restart); trim to fit.
        self._evict_over_cap()


def _encode(item: Mapping[str, Any]) -> bytes:
    return json.dumps(item, default=str).encode("utf-8")
